File: backend/src/main/java/com/smartroute/smartroute1/service/RouteService.java
package com.smartroute.smartroute1.service;

import com.smartroute.smartroute1.endpoint.dto.RouteDto;
import com.smartroute.smartroute1.endpoint.dto.SaveRouteDto;
import com.smartroute.smartroute1.endpoint.dto.ShareRouteDto;
import com.smartroute.smartroute1.endpoint.dto.ViewRouteDto;
import com.smartroute.smartroute1.entity.ApplicationUser;

import java.util.List;

public interface RouteService {
    /**
     * Save route for the corresponding user.
     *
     * @param route the routedto to save
     * @param user  the user it belongs to
     */
    void saveRoute(SaveRouteDto route, ApplicationUser user);

    /**
     * Get all routes saved routes from a user.
     *
     * @param user the user to get the routes from
     * @return the view route dto
     */
    List<ViewRouteDto> getRoutes(ApplicationUser user);

    /**
     * Get one route by id.
     *
     * @param id    the id
     * @param email to verify email
     * @return the ViewRouteDto
     */
    ViewRouteDto getRoute(Long id, String email);

    /**
     * Delete one by id.
     *
     * @param id    the id
     * @param email email of user this belongs to
     */
    void deleteRoute(Long id, String email);


    /**
     * Shares a route with other users.
     *
     * @param id the id of the route
     * @param email1 the email of the owner of the route
     * @param firends the emails of the friends we share the route with
     */
    void addShare(Long id, String email1, ShareRouteDto firends);
}

File: frontend/src/services/route.service.ts
import {HttpClient} from "@angular/common/http";
import {inject, Injectable} from "@angular/core";
import {BehaviorSubject, Observable} from "rxjs";
import {Globals} from "src/global/globals";
import {GeneratedRouteDto} from "src/app/dtos/route";
import {SaveRouteDto, ViewRouteDto} from "../app/dtos/recommended-activity";

@Injectable({
  providedIn: 'root',
})
export class RouteService {
  private readonly httpClient: HttpClient = inject(HttpClient);
  private readonly globals: Globals = inject(Globals);

  private readonly routeBaseUri: string = this.globals.backendUri + '/route';

  private lastRoute: GeneratedRouteDto | null = null;

  /**
   * Generate route for ('lat', 'long') with length 'length'.
   */
  getGeneratedRoute(lat: number, long: number, length: number): Observable<GeneratedRouteDto> {
    let route = this.httpClient.get<GeneratedRouteDto>(this.routeBaseUri, {
      params: {
        "lat": lat,
        "long": long,
        "length": length
      }
    });
    return route;
  }

  getLastGeneratedRoute(): GeneratedRouteDto | null {
    return this.lastRoute;
  }

  /**
   * Save Route to user
   * @param dto
   */
  saveRoute(dto: SaveRouteDto) {
    return this.httpClient.post(this.routeBaseUri + '/save', dto);
  }

  /**
   * Get one route by id
   * @param id
   */
  getRoute(id: number): Observable<ViewRouteDto> {
    return this.httpClient.get<ViewRouteDto>(`${this.routeBaseUri}/${id}`);
  }

  /**
   * Get all saved routes from the user
   */
  getRoutes(): Observable<ViewRouteDto[]> {
    return this.httpClient.get<ViewRouteDto[]>(this.routeBaseUri + '/get');
  }

  /**
   * delete route with id
   * @param id
   */
  delete(id: number) {
    return this.httpClient.delete(this.routeBaseUri + `/${id}`);
  }

  share(id: number, friend: string[]){
    console.log(this.routeBaseUri + `/${id}/share`, {"friends":friend});
    return this.httpClient.put(this.routeBaseUri + `/${id}/share`, {"friends":friend})
  }
}
